staging/src/k8s.io/apiserver/pkg/endpoints/request/context.go

```diff
@@ -94,3 +98,49 @@ func AuditEventFrom(ctx context.Context) *audit.Event {
 	ev, _ := ctx.Value(auditKey).(*audit.Event)
 	return ev
 }
+
+func AddEarlyAuditAnnotation(parent context.Context, key, value string) context.Context {
+	if ae := AuditEventFrom(parent); ae != nil {
+		panic("invalid attempt to set early audit annotation on context after WithAuditEvent has been called")
+	}
+
+	annotations, ok := parent.Value(auditAnnotationsKey).(map[string]string)
+
+	if v, hasKey := annotations[key]; hasKey && v != value {
+		klog.Warningf("failed to set early audit annotation %q to %q, it has already been set to %q", key, value, v)
+		return parent
+	}
+
+	// allocate a new map if needed only after we confirm our intent to write the new key
+	if !ok {
+		annotations = map[string]string{}
+	}
+
+	annotations[key] = value
+
+	// avoid extra wrapping when we mutated a pre-existing map
+	if ok {
+		return parent
+	}
+
+	return WithValue(parent, auditAnnotationsKey, annotations)
+}
+
+func EarlyAuditAnnotationsFrom(ctx context.Context) map[string]string {
+	if ae := AuditEventFrom(ctx); ae != nil {
+		panic("invalid attempt to get early audit annotations from context after WithAuditEvent has been called")
+	}
+
+	annotations, ok := ctx.Value(auditAnnotationsKey).(map[string]string)
+	if !ok {
+		return nil // return early to avoid allocations
+	}
+
+	// prevent mutation of stored map as that is only allowed via AddEarlyAuditAnnotation
+	annotationsCopy := make(map[string]string, len(annotations))
```

Comment:
Both the copy and the map make this look pretty expensive then we end up copying again. How would a `[]struct{k,v string}` look here?